scripts/src/java/org/oppia/android/scripts/apkstats/ApkAnalyzerClient.kt

```diff
@@ -0,0 +1,105 @@
+  fun computeFileSize(inputApkPath: String): Long =
+    apkSizeCalculator.getFullApkRawSize(File(inputApkPath).toPath())
+
+  /**
+   * Returns the estimated download size of the specified APK as similarly reported by apkanalyzer.
+   */
+  fun computeDownloadSize(inputApkPath: String): Long =
+    apkSizeCalculator.getFullApkDownloadSize(File(inputApkPath).toPath())
+
+  /**
+   * Returns the list of required features of the specified APK as similarly reported by
+   * apkanalyzer.
+   */
+  fun computeFeatures(inputApkPath: String): List<String> {
+    val apkInfo = AndroidApplicationInfo.parseBadging(aapt2Client.dumpBadging(inputApkPath))
+    return apkInfo.usesFeature.keys.toList()
+  }
+
+  /**
+   * Returns a full comparison of the two specified APKs, similar to apkanalyzer. Note that each
+   * entry of the returned list represents a single line (the list is in the order of the comparison
+   * output).
+   */
+  fun compare(inputApkPath1: String, inputApkPath2: String): List<String> {
+    return Archives.open(File(inputApkPath1).toPath()).use { apkArchive1 ->
+      Archives.open(File(inputApkPath2).toPath()).use { apkArchive2 ->
+        walkParseTree(ApkFileByFileDiffParser.createTreeNode(apkArchive1, apkArchive2))
+      }
+    }
+  }
+
+  // Based on the apkanalyzer CLI implementation.
+  private fun walkParseTree(node: DefaultMutableTreeNode?): List<String> {
+    return node?.let {
+      (node.userObject as? ApkDiffEntry)?.let { entry ->
+        val path = entry.path.toString()
+        when {
+          // Encountered a folder. Recurse down it in case its sizes aren't accurate (descendants
+          // may have changed).
+          path.endsWith("/") -> {
+            (0 until node.getChildCount())
+              .map { node.getChildAt(it) }
+              .flatMap { walkParseTree(it as? DefaultMutableTreeNode) }
+          }
+          // Only record the entry if there's a difference.
+          entry.oldSize != entry.newSize -> {
+            listOf("${entry.oldSize}\t${entry.newSize}\t${entry.size}\t${entry.path}")
+          }
+          else -> null
+        }
+      }
+    } ?: listOf()
+  }
+
+  /**
+   * Returns the map of dex files to method counts contained within the specified APK file, as
+   * similarly reported by apkanalyzer.
+   */
+  fun computeDexReferencesList(inputApkPath: String): Map<String, Int> {
+    return collectZipEntries(inputApkPath) {
+      it.endsWith(".dex")
+    }.mapValues { (_, dexFileContents) ->
+      DexFileStats.create(listOf(DexFiles.getDexFile(dexFileContents))).referencedMethodCount
+    }
+  }
+
+  private fun collectZipEntries(
+    inputZipFile: String,
+    predicate: (String) -> Boolean
+  ): Map<String, ByteArray> {
+    return ZipFile(inputZipFile).use { zipFile ->
+      zipFile.entries()
+        .asSequence()
+        .filter { predicate(it.name) }
+        .associateBy { it.name }
+        .mapValues { (_, entry) -> zipFile.getInputStream(entry).readBytes() }
```

Comment:
Close stream.

Reply:
Doesn't actually need to be closed since it's part of a ZipFile class that's already being closed (per ZipFile docs).